pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,338 @@
+	ctx = logging.WithLogger(ctx, logging.FromContext(ctx).Named("nodemetrics").With("node", req.Name))
+	// Retrieve node from reconcile request
+	node := &v1.Node{}
+	if err := c.KubeClient.Get(ctx, req.NamespacedName, node); err != nil {
+		if errors.IsNotFound(err) {
+			// Remove gauge due to node deletion
+			if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+				c.deleteGaguges(labels)
+			} else {
+				logging.FromContext(ctx).Errorf("Failed to delete gauge: failed to locate labels")
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after node labels are updated
+	if labels, ok := c.LabelsMap[req.NamespacedName]; ok {
+		c.deleteGaguges(labels)
+	}
+	newlabels, err := c.generateLabels(node)
+	if err != nil {
+		logging.FromContext(ctx).Errorf("Failed to generate labels: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	if err := c.updateGaguges(ctx, node, newlabels); err != nil {
+		logging.FromContext(ctx).Errorf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+	c.LabelsMap[req.NamespacedName] = newlabels
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Errorf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		WithOptions(controller.Options{MaxConcurrentReconciles: 10000}).
+		Complete(c)
+	return err
+}
+
+func (c *Controller) deleteGaguges(labels *prometheus.Labels) {
+	for _, gaugeVec := range c.GaugeVecMap {
+		gaugeVec.Delete(*labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node) (*prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
```

Comment:
naming nit: 
we're already contextualized with the "metric" concept by nature of being in the metrics controller. You could just say `labels := prometheus.Labels{}` and it would read just as well.

Reply:
I was thinking about distinguish the k8s labels and the prometheus labels. Maybe I should name them prometheusLabels?

Reply:
This is fine.